File: src/decentralized/golang/dragonfly/dragonfly.go
package dragonfly

import (
	"context"
	"crypto/sha256"
	"encoding/hex"
	"errors"
	"fmt"
	"io"
	"net/http"
	"sync"
	"time"
)

type Chunk struct {
	ContentID string
	Index     int
	Data      []byte
	Hash      string
}

type OriginFetcher struct {
	gatewayURL string
	client     *http.Client
}

func NewOriginFetcher(gatewayURL string) *OriginFetcher {
	return &OriginFetcher{
		gatewayURL: gatewayURL,
		client:     &http.Client{Timeout: 30 * time.Second},
	}
}

func (f *OriginFetcher) FetchRange(ctx context.Context, assetID string, start, end int64) ([]byte, error) {
	url := fmt.Sprintf("%s/object/%s", f.gatewayURL, assetID)
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
	if err != nil {
		return nil, err
	}
	req.Header.Set("Range", fmt.Sprintf("bytes=%d-%d", start, end))
	resp, err := f.client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	return io.ReadAll(resp.Body)
}

type DiskCache struct {
	mu      sync.RWMutex
	store   map[string][]byte
	maxSize int64
	used    int64
}

func NewDiskCache(maxSize int64) *DiskCache {
	return &DiskCache{store: make(map[string][]byte), maxSize: maxSize}
}

func (dc *DiskCache) Store(key string, data []byte) {
	dc.mu.Lock()
	defer dc.mu.Unlock()
	if dc.used+int64(len(data)) > dc.maxSize {
		dc.evict(int64(len(data)))
	}
	dc.store[key] = data
	dc.used += int64(len(data))
}

func (dc *DiskCache) Get(key string) ([]byte, bool) {
	dc.mu.RLock()
	defer dc.mu.RUnlock()
	d, ok := dc.store[key]
	return d, ok
}

func (dc *DiskCache) Evict(key string) {
	dc.mu.Lock()
	if d, ok := dc.store[key]; ok {
		dc.used -= int64(len(d))
		delete(dc.store, key)
	}
	dc.mu.Unlock()
}

func (dc *DiskCache) evict(needed int64) {
	for k, v := range dc.store {
		delete(dc.store, k)
		dc.used -= int64(len(v))
		if dc.maxSize-dc.used >= needed {
			break
		}
	}
}

type PieceServer struct {
	cache *DiskCache
	mux   *http.ServeMux
}

func NewPieceServer(cache *DiskCache) *PieceServer {
	ps := &PieceServer{cache: cache, mux: http.NewServeMux()}
	ps.mux.HandleFunc("/piece/", ps.handlePiece)
	return ps
}

func (ps *PieceServer) ServeHTTP(w http.ResponseWriter, r *http.Request) {
	ps.mux.ServeHTTP(w, r)
}

func (ps *PieceServer) handlePiece(w http.ResponseWriter, r *http.Request) {
	key := r.URL.Path[len("/piece/"):]
	data, ok := ps.cache.Get(key)
	if !ok {
		http.NotFound(w, r)
		return
	}
	w.Header().Set("Content-Length", fmt.Sprintf("%d", len(data)))
	w.Write(data)
}

type CacheGarbageCollector struct {
	cache    *DiskCache
	interval time.Duration
	policy   GCPolicy
}

type GCPolicy struct {
	MaxAge    time.Duration
	MaxSize   int64
	MinFree   int64
}

func NewCacheGarbageCollector(cache *DiskCache, interval time.Duration, policy GCPolicy) *CacheGarbageCollector {
	return &CacheGarbageCollector{cache: cache, interval: interval, policy: policy}
}

func (gc *CacheGarbageCollector) Run(ctx context.Context) {
	ticker := time.NewTicker(gc.interval)
	defer ticker.Stop()
	for {
		select {
		case <-ctx.Done():
			return
		case <-ticker.C:
			gc.collect()
		}
	}
}

func (gc *CacheGarbageCollector) collect() {
	gc.cache.mu.Lock()
	defer gc.cache.mu.Unlock()
	if gc.cache.used < gc.policy.MaxSize {
		return
	}
	excess := gc.cache.used - gc.policy.MaxSize + gc.policy.MinFree
	for k, v := range gc.cache.store {
		if excess <= 0 {
			break
		}
		delete(gc.cache.store, k)
		excess -= int64(len(v))
		gc.cache.used -= int64(len(v))
	}
}

type IntegrityVerifier struct{}

func NewIntegrityVerifier() *IntegrityVerifier {
	return &IntegrityVerifier{}
}

func (iv *IntegrityVerifier) Verify(chunk Chunk) error {
	sum := sha256.Sum256(chunk.Data)
	actual := hex.EncodeToString(sum[:])
	if actual != chunk.Hash {
		return errors.New("chunk hash mismatch")
	}
	return nil
}

type LocalCache struct {
	mu    sync.RWMutex
	store map[string]Chunk
}

func NewLocalCache() *LocalCache {
	return &LocalCache{store: make(map[string]Chunk)}
}

func (lc *LocalCache) Write(chunk Chunk) {
	lc.mu.Lock()
	lc.store[chunkKey(chunk)] = chunk
	lc.mu.Unlock()
}

func (lc *LocalCache) Read(contentID string, index int) (Chunk, bool) {
	lc.mu.RLock()
	defer lc.mu.RUnlock()
	c, ok := lc.store[contentID+":"+fmt.Sprintf("%d", index)]
	return c, ok
}

func chunkKey(c Chunk) string {
	return c.ContentID + ":" + fmt.Sprintf("%d", c.Index)
}

type PeerSource interface {
	FetchChunk(ctx context.Context, contentID string, index int) (Chunk, error)
}

type PieceDownloader struct {
	peers      []PeerSource
	verifier   *IntegrityVerifier
	localCache *LocalCache
}

func NewPieceDownloader(peers []PeerSource, verifier *IntegrityVerifier, lc *LocalCache) *PieceDownloader {
	return &PieceDownloader{peers: peers, verifier: verifier, localCache: lc}
}

func (pd *PieceDownloader) Download(ctx context.Context, contentID string, index int) (Chunk, error) {
	if chunk, ok := pd.localCache.Read(contentID, index); ok {
		return chunk, nil
	}
	for _, peer := range pd.peers {
		chunk, err := peer.FetchChunk(ctx, contentID, index)
		if err != nil {
			continue
		}
		if err := pd.verifier.Verify(chunk); err != nil {
			continue
		}
		pd.localCache.Write(chunk)
		return chunk, nil
	}
	return Chunk{}, errors.New("no peer could serve chunk")
}

type PieceUploader struct {
	peers []PeerSink
}

type PeerSink interface {
	SendChunk(ctx context.Context, chunk Chunk) error
	ID() string
}

func NewPieceUploader(peers []PeerSink) *PieceUploader {
	return &PieceUploader{peers: peers}
}

func (pu *PieceUploader) Redistribute(ctx context.Context, chunk Chunk) {
	for _, peer := range pu.peers {
		go func(p PeerSink) {
			p.SendChunk(ctx, chunk)
		}(peer)
	}
}

type RetryEngine struct {
	downloader *PieceDownloader
	maxRetries int
	backoff    time.Duration
}

func NewRetryEngine(dl *PieceDownloader, maxRetries int, backoff time.Duration) *RetryEngine {
	return &RetryEngine{downloader: dl, maxRetries: maxRetries, backoff: backoff}
}

func (re *RetryEngine) Download(ctx context.Context, contentID string, index int) (Chunk, error) {
	var lastErr error
	for i := 0; i < re.maxRetries; i++ {
		chunk, err := re.downloader.Download(ctx, contentID, index)
		if err == nil {
			return chunk, nil
		}
		lastErr = err
		select {
		case <-ctx.Done():
			return Chunk{}, ctx.Err()
		case <-time.After(re.backoff * time.Duration(i+1)):
		}
	}
	return Chunk{}, fmt.Errorf("max retries exceeded: %w", lastErr)
}

File: src/decentralized/python/dragonfly/dragonfly.py
from __future__ import annotations
import hashlib
import threading
import time
import urllib.request
from dataclasses import dataclass, field
from typing import Any, Dict, List, Optional, Protocol, Tuple


@dataclass
class Chunk:
    content_id: str
    index: int
    data: bytes
    hash: str


class OriginFetcher:
    def __init__(self, gateway_url: str):
        self._gateway_url = gateway_url

    def fetch_range(self, asset_id: str, start: int, end: int) -> bytes:
        url = f"{self._gateway_url}/object/{asset_id}"
        req = urllib.request.Request(url, headers={"Range": f"bytes={start}-{end}"})
        with urllib.request.urlopen(req, timeout=30) as resp:
            return resp.read()


class DiskCache:
    def __init__(self, max_size: int = 10 * 1024 * 1024 * 1024):
        self._lock = threading.RLock()
        self._store: Dict[str, bytes] = {}
        self._max_size = max_size
        self._used = 0

    def store(self, key: str, data: bytes):
        with self._lock:
            if self._used + len(data) > self._max_size:
                self._evict(len(data))
            self._store[key] = data
            self._used += len(data)

    def get(self, key: str) -> Optional[bytes]:
        with self._lock:
            return self._store.get(key)

    def evict(self, key: str):
        with self._lock:
            data = self._store.pop(key, None)
            if data:
                self._used -= len(data)

    def _evict(self, needed: int):
        for k in list(self._store.keys()):
            if self._max_size - self._used >= needed:
                break
            data = self._store.pop(k)
            self._used -= len(data)


class PieceServer:
    def __init__(self, cache: DiskCache):
        self._cache = cache

    def serve(self, key: str) -> Optional[bytes]:
        return self._cache.get(key)

    def make_wsgi_app(self):
        cache = self._cache

        def app(environ, start_response):
            path = environ.get("PATH_INFO", "")
            key = path.lstrip("/piece/")
            data = cache.get(key)
            if data is None:
                start_response("404 Not Found", [])
                return [b"not found"]
            start_response("200 OK", [("Content-Length", str(len(data)))])
            return [data]

        return app


class CacheGarbageCollector:
    def __init__(self, cache: DiskCache, interval: float = 60.0, max_size: int = 10 * 1024 * 1024 * 1024):
        self._cache = cache
        self._interval = interval
        self._max_size = max_size

    def run(self):
        while True:
            time.sleep(self._interval)
            self._collect()

    def _collect(self):
        with self._cache._lock:
            if self._cache._used < self._max_size:
                return
            excess = self._cache._used - self._max_size
            for k in list(self._cache._store.keys()):
                if excess <= 0:
                    break
                data = self._cache._store.pop(k)
                excess -= len(data)
                self._cache._used -= len(data)


class IntegrityVerifier:
    def verify(self, chunk: Chunk) -> bool:
        actual = hashlib.sha256(chunk.data).hexdigest()
        return actual == chunk.hash


class LocalCache:
    def __init__(self):
        self._lock = threading.RLock()
        self._store: Dict[str, Chunk] = {}

    def write(self, chunk: Chunk):
        key = f"{chunk.content_id}:{chunk.index}"
        with self._lock:
            self._store[key] = chunk

    def read(self, content_id: str, index: int) -> Optional[Chunk]:
        key = f"{content_id}:{index}"
        with self._lock:
            return self._store.get(key)


class PeerSourceProtocol(Protocol):
    def fetch_chunk(self, content_id: str, index: int) -> Chunk: ...


class PieceDownloader:
    def __init__(self, peers: List[PeerSourceProtocol], verifier: IntegrityVerifier, local_cache: LocalCache):
        self._peers = peers
        self._verifier = verifier
        self._local_cache = local_cache

    def download(self, content_id: str, index: int) -> Chunk:
        cached = self._local_cache.read(content_id, index)
        if cached:
            return cached
        for peer in self._peers:
            try:
                chunk = peer.fetch_chunk(content_id, index)
                if self._verifier.verify(chunk):
                    self._local_cache.write(chunk)
                    return chunk
            except Exception:
                continue
        raise RuntimeError(f"no peer could serve chunk {content_id}:{index}")


class PeerSinkProtocol(Protocol):
    def send_chunk(self, chunk: Chunk): ...
    def id(self) -> str: ...


class PieceUploader:
    def __init__(self, peers: List[PeerSinkProtocol]):
        self._peers = peers

    def redistribute(self, chunk: Chunk):
        for peer in self._peers:
            t = threading.Thread(target=self._send, args=(peer, chunk), daemon=True)
            t.start()

    def _send(self, peer: PeerSinkProtocol, chunk: Chunk):
        try:
            peer.send_chunk(chunk)
        except Exception:
            pass


class RetryEngine:
    def __init__(self, downloader: PieceDownloader, max_retries: int = 3, backoff: float = 1.0):
        self._downloader = downloader
        self._max_retries = max_retries
        self._backoff = backoff

    def download(self, content_id: str, index: int) -> Chunk:
        last_err = None
        for i in range(self._max_retries):
            try:
                return self._downloader.download(content_id, index)
            except Exception as e:
                last_err = e
                time.sleep(self._backoff * (i + 1))
        raise RuntimeError(f"max retries exceeded: {last_err}")
